Fix "key_as_string" for date histogram and epoch_millis/epoch_second format with time zone

cbuescher opened the pull request:
When doing a `date_histogram` aggregation with `"format":"epoch_millis"` or `"format" : "epoch_second"` and using a time zone other than UTC, the `key_as_string` ouput in the response does not reflect the UTC timestamp that is used as the key. This happens because when applying the `time_zone` in DocValueFormat.DateTime to an epoch-based formatter, this adds the time zone offset to the value being formated. Instead we should adjust the added display offset to get back the utc instance in EpochTimePrinter.

Closes #19038

jpountz commented on core/src/main/java/org/elasticsearch/search/DocValueFormat.java:
```diff
@@ -124,14 +126,20 @@ public BytesRef parseBytesRef(String value) {
             this.timeZone = Objects.requireNonNull(timeZone);
             this.parser = new DateMathParser(formatter);
+            if (formatter.format().startsWith("epoch_millis") || formatter.format().startsWith("epoch_second")) {
+                this.formatter = formatter.printer();
```
do we really need a special case here or could we fix the printer impl for these formats to apply the timezone when printing a date?

cbuescher replied:
I looked into this some more. It's problematic, the actual joda EpochPrinter prints whatever time instant (long) is given to it, the time zone adjustment happens in joda DateTimeFormatter.printTo(Appendable, long, Chronology), and if we use `withZone()` (which we need for other formatters to work correct) the time zone offset gets added. I don't see another way around this atm. but am open to suggestions where to dig further if you have ideas.

jpountz replied:
I'm wondering that this might be something that should be fixed upstream?

cbuescher replied:
Sorry, I was mistaken when I thought EpochTimePrinter is part of Joda, its in our code base and we can fix the problem there. I will rebase and update this PR.

cbuescher commented:
@jpountz sorry for the confusion about EpochTimePrinter, this is actually part of our codebase and we can correct the time zone adjustment directly there. I'm not completely sure if `epoch_millis` format is supposed to always output the date instance in UTC, maybe @spinscale can conform this since it look like he was the one that introduced the printer class.

spinscale commented:
yes, epoch time is considered the seconds elapsed since 1970 in UTC timezone.

jpountz commented:
+1 to always UTC

jpountz commented on core/src/main/java/org/elasticsearch/common/joda/Joda.java:
```diff
@@ -378,18 +377,18 @@ public int estimatePrintedLength() {
         public void printTo(StringBuffer buf, long instant, Chronology chrono, int displayOffset, DateTimeZone displayZone, Locale locale) {
             if (hasMilliSecondPrecision) {
-                buf.append(instant);
+                buf.append(instant - displayOffset);
```
maybe leave a comment where this comes from?

jpountz commented:
LGTM

cbuescher committed:
Fix key_as_string for date histogram and epoch_millis/epoch_second format

When doing a `date_histogram` aggregation with `"format":"epoch_millis"` or
`"format" : "epoch_second"` and using a time zone other than UTC, the
`key_as_string` ouput in the response does not reflect the UTC timestamp that is
used as the key. This happens because when applying the `time_zone` in
DocValueFormat.DateTime to an epoch-based formatter, this adds the time zone
offset to the value being formated. Instead we should adjust the added display
offset to get back the utc instance in EpochTimePrinter.

Closes #19038

cbuescher commented:
Merged to 2.4 with 5301ee30
